### DISCUSSION TITLE
How to extract pores from a 3D image

### DISCUSSION THREAD

**Discussion by UnaisAshraf** [PMEAL/OpenPNM#2767] [Q&A]
Dear professor,

I want to extract pores and throats from a 3D-image. It is stl file with a size of 1gb. I don't know how i can import this stl file in openPNM.

![3D image](https://github.com/PMEAL/OpenPNM/assets/132771045/cd12245b-a2cf-4be9-966b-14dc08d1dc49)

This is my image and I want to extract the pores and throats from here to something like this below.

![pore network](https://github.com/PMEAL/OpenPNM/assets/132771045/aac6bfe6-927d-45f3-b1f3-413c2d487557)

Can anyone tell me how i can do this. I would really appreciate your help.

**Comment by likeduck** [comment]
Hi! @UnaisAshraf 
  If you want to extracted PNM using SNOW algorithm, you need change stl to tiff in order to read data into code. SNOW algorithn is [here](https://porespy.org/examples/networks/tutorials/snow_advanced.html).

**Reply by UnaisAshraf** [reply]
but is it possible to convert 1gb size of stl file into tiff file?

**Reply by jgostick** [reply]
Whatever software you are using to view the stl file should have a saveas tiff feature.  Otherwise, you'll have to ask chatgtp.  You could also convert to a numpy array uf that helps.

**Reply by UnaisAshraf** [reply]
so i converted my stl file to tiff format like you said and used the same snow extraction (advanced) code but i am getting it like this when i run it in paraview

![image](https://github.com/PMEAL/OpenPNM/assets/132771045/c6801d6f-99ac-4511-8336-3593eb05ee95)
 its like its in 2D. Can you tell me how i can fix this. what i want is something like this below image

![image](https://github.com/PMEAL/OpenPNM/assets/132771045/c90e9f40-6d6d-49b9-b48a-18335836f03a)

i used the same code given in the example only changed im to my own image since i am just a beginner in openPNM.

import imageio
import numpy as np
import porespy as ps
import openpnm as op
import scipy.ndimage as spim
import matplotlib.pyplot as plt
from porespy.filters import find_peaks, trim_saddle_points, trim_nearby_peaks
from porespy.tools import randomize_colors
from skimage.segmentation import watershed

ps.settings.tqdm['disable'] = True
ps.visualization.set_mpl_style()
np.random.seed(10)

im = imageio.v2.imread('C:/Users/unais/Downloads/3dpnm.tiff')
fig, ax = plt.subplots()
ax.imshow(ps.visualization.show_planes(im), cmap=plt.cm.bone);

sigma = 0.8
dt = spim.distance_transform_edt(input=im)
dt1 = spim.gaussian_filter(input=dt, sigma=sigma)
peaks = find_peaks(dt=dt)

print('Initial number of peaks: ', spim.label(peaks)[1])
peaks = trim_saddle_points(peaks=peaks, dt=dt1)
print('Peaks after trimming saddle points: ', spim.label(peaks)[1])
peaks = trim_nearby_peaks(peaks=peaks, dt=dt)
peaks, N = spim.label(peaks)
print('Peaks after trimming nearby peaks: ', N)

regions = watershed(image=-dt, markers=peaks, mask=dt > 0)
regions = randomize_colors(regions)

fig, ax = plt.subplots()
ax.imshow((regions*im), cmap=plt.cm.nipy_spectral);

net = ps.networks.regions_to_network(regions*im, voxel_size=1) 
pn = op.io.network_from_porespy(net)
print(pn)

im = ps.tools.align_image_with_openpnm(im)
imageio.volsave('image.tif', np.array(im, dtype=np.int8))
op.io.project_to_vtk(project=pn.project)

**Reply by likeduck** [reply]
It must be ensured that the tiff image to be extracted is three-dimensional. You can open it in paraview in advance to check if it is three-dimensional or or view it through your first fig function.
![Figure 1](https://github.com/PMEAL/OpenPNM/assets/117707506/95febbb7-8e59-4196-b0d1-2db7259f09bf)
It has three cross-section and PNM is:
![Figure 4](https://github.com/PMEAL/OpenPNM/assets/117707506/2fee7b45-bcde-4944-999f-696ac4d049a3)

**Reply by likeduck** [reply]
Or you can replace the code with：
```
# net = ps.networks.regions_to_network(regions*im, voxel_size=1)
net = ps.networks.snow2(im, voxel_size=1, boundary_width=0)
# pn = op.io.network_from_porespy(net)
pn = op.io.network_from_porespy(net.network)
```

**Reply by beidellite** [reply]
Hello 

If you use python there is a nifty package stltovoxel which will convert your stl file to a 3D stack, you can define the resolution, voxel size or colors.
https://pypi.org/project/stl-to-voxel/

Additionally if you are in academia and have access to a windows robust desktop, you may obtain and download Dragonfly ORS software, apart from having nifty modules, it has an openpnm utility that you can use to generate the stick and ball model, might be worth it to give it a  look. But I think the stl-to-voxel package will/should get you what you need.

**Comment by ZhouChuanyou** [comment]
![net](https://github.com/user-attachments/assets/533564fe-5837-43d0-a00f-580a840fa049)
Using the SNOW algorithm can extract the pore network like the picture displayed. 
I want to export the pore network to Statoil format (_link1.dat, _link2.dat, _node1.dat and _node2.dat). 
Is there any package to do it? Thank you!

**Reply by jgostick** [reply]
We used to export to statoil in V2, but decided not to port it over to V3 since it didn't really make sense as there are other better formats.  For instance to use the network in Avizo you can use `op.io.network_to_pergeos`.

**Reply by ZhouChuanyou** [reply]
Thank you very much for your kind reply and also the useful suggestion!
